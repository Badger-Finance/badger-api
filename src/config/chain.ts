import { ethers } from 'ethers';
import { SettDefinition } from '../interface/Sett';
import { Token } from '../interface/Token';
import { Provider } from './constants';
import { ethSetts } from './setts/eth-setts';

export interface Chain {
  readonly name: string;
  readonly symbol: string;
  readonly chainId: string;
  readonly tokens: Token[];
  readonly setts: SettDefinition[];
  readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
  public name = 'Ethereum';
  public symbol = 'eth';
  public chainId = '0x01';
  public tokens = [];
  public setts = ethSetts;
  public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
  public name = 'BinanceSmartChain';
  public symbol = 'bsc';
  public chainId = '0x61';
  public tokens = [];
  public setts = [];
  public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 * TODO: Should not supported throw a bad request instead?
 */
export const resolveChainQuery = (symbol: string): Chain => {
  if (!symbol) return eth;
  const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
  if (!chain) return eth;
  return chain;
};
