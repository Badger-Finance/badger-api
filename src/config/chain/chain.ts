import { BadRequest } from '@tsed/exceptions';
import { ethers } from 'ethers';
import { SettDefinition } from '../../interface/Sett';
import { Token } from '../../interface/Token';
import { Provider } from '../constants';
import { bscSetts } from './bsc';
import { ethSetts } from './eth';

export interface Chain {
	readonly name: string;
	readonly symbol: string;
	readonly chainId: string;
	readonly tokens: Token[];
	readonly setts: SettDefinition[];
	readonly provider: ethers.providers.JsonRpcProvider;
}

export class Ethereum implements Chain {
	public name = 'Ethereum';
	public symbol = 'eth';
	public chainId = '0x01';
	public tokens = [];
	public setts = ethSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Cloudflare);
}

export class BinanceSmartChain implements Chain {
	public name = 'BinanceSmartChain';
	public symbol = 'bsc';
	public chainId = '0x61';
	public tokens = [];
	public setts = bscSetts;
	public provider = new ethers.providers.JsonRpcProvider(Provider.Binance);
}

export const eth = new Ethereum();
export const bsc = new BinanceSmartChain();

export const supportedChains = [eth, bsc];

/**
 * Resolve a request query param chain lookup.
 * Failure to specify a chain, or requesting a non-supported
 * chain will default to Ethereum.
 */
export const resolveChainQuery = (symbol: string): Chain => {
	if (!symbol) return eth;
	const chain = supportedChains.find((c) => c.symbol.toLowerCase() === symbol.toLowerCase());
	if (!chain) throw new BadRequest('Invalid chain');
	return chain;
};
