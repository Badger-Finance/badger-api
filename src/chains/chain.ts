import { Arbitrum } from './config/arbitrum.config';
import { BinanceSmartChain } from './config/bsc.config';
import { Chain } from './config/chain.config';
import { Ethereum } from './config/eth.config';
import { Polygon } from './config/matic.config';

/**
 * Instantiate objects for registration.
 */
export const loadChains = (): Chain[] => {
  return [new Ethereum(), new BinanceSmartChain(), new Polygon(), new Arbitrum()];
};
