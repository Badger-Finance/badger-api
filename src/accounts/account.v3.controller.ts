import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { AccountsService } from './accounts.service';
import { AccountModel } from './interfaces/account-model.interface';
import { QueryParamError } from '../errors/validation/query.param.error';

@Controller('/account')
export class AccountV3Controller {
  @Inject()
  accountsService!: AccountsService;

  @Get()
  @ContentType('json')
  @Summary('Get badger user account information')
  @Description(
    'Return key user information for a given account. Includes positions, earnings from use, and claimable balances.',
  )
  @Returns(200, AccountModel)
  @Returns(400).Description('Not a valid chain')
  @Returns(404).Description('Not a valid account')
  async getAccount(
    @QueryParams('accountId') userId: string,
    @QueryParams('chain') chain: Network,
  ): Promise<AccountModel> {
    if (!userId) throw new QueryParamError('userId');

    return this.accountsService.getAccount(Chain.getChain(chain), userId);
  }
}
