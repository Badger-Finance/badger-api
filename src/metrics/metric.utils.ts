import { getAccounts } from '../accounts/accounts.utils';
import { loadChains } from '../chains/chain';
import { Chain } from '../chains/config/chain.config';
import { ProtocolSummary } from '../protocols/interfaces/protocol-summary.interface';
import { getCachedVault } from '../vaults/vaults.utils';
import { ProtocolMetrics, ProtocolSettsMetrics } from './interfaces/metrics.interface';

export const getProtocolMetrics = async (): Promise<ProtocolMetrics> => {
  const [totalUsers, settMetrics] = await Promise.all([getProtocolTotalUsers(), getProtocolSettMetrics()]);
  return { ...settMetrics, totalUsers };
};

export async function getProtocolTotalUsers(): Promise<number> {
  const chains = loadChains();
  const usersAcrossChains = await Promise.all(chains.map((chain) => getAccounts(chain)));
  return new Set([...usersAcrossChains.flat()]).size;
}

export async function getProtocolSettMetrics(): Promise<ProtocolSettsMetrics> {
  const chains = loadChains();
  const multichainsSummary = await Promise.all(chains.map((chain) => getChainMetrics(chain)));

  let totalValueLocked = 0;
  let totalVaults = 0;

  for (const protocolSummary of multichainsSummary) {
    totalValueLocked += protocolSummary.totalValue;
    totalVaults += protocolSummary.setts?.length ?? 0;
  }

  return { totalValueLocked, totalVaults };
}

export async function getChainMetrics(chain: Chain): Promise<ProtocolSummary> {
  const setts = await Promise.all(chain.setts.map((sett) => getCachedVault(sett)));
  const totalValue = setts.reduce((total, sett) => total + sett.value, 0);

  return { totalValue, setts: setts.map(({ name, balance, value }) => ({ name, balance, value })) };
}
