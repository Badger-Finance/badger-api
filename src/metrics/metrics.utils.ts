import { ProtocolMetrics } from '@badger-dao/sdk';
import { NotFound } from '@tsed/exceptions';

import { getAccounts } from '../accounts/accounts.utils';
import { getDataMapper } from '../aws/dynamodb.utils';
import { ProtocolMetricSnapshot } from '../aws/models/protocol-metric-snapshot.model';
import { Chain } from '../chains/config/chain.config';
import { getCachedVault } from '../vaults/vaults.utils';
import { MetricType } from './enums/metric-type';

export async function getChainMetrics(chains: Chain[]): Promise<ProtocolMetrics> {
  let totalUsers = 0;
  let totalValueLocked = 0;
  let totalVaults = 0;

  for (const chain of chains) {
    const accounts = await getAccounts(chain);
    totalUsers += accounts.length;

    const vaults = await chain.vaults.all();
    const vaultSummaries = await Promise.all(
      vaults.map(async (vault) => {
        const { balance, value } = await getCachedVault(chain, vault);
        return { name: vault.name, balance, value };
      }),
    );
    const totalValue = vaultSummaries.reduce((total, vault) => total + vault.value, 0);

    totalValueLocked += totalValue;
    totalVaults += vaultSummaries.length;
  }

  return { totalValueLocked, totalVaults, totalUsers };
}

export async function queryProtocolMetrics(): Promise<ProtocolMetrics> {
  const errorMessage = 'Protocol metrics not available';
  try {
    const mapper = getDataMapper();
    for await (const metric of mapper.query(
      ProtocolMetricSnapshot,
      { type: MetricType.Protocol },
      { scanIndexForward: false, limit: 1 },
    )) {
      return metric;
    }
    throw new NotFound(errorMessage);
  } catch (error) {
    console.error(error);
    throw new NotFound(errorMessage);
  }
}
