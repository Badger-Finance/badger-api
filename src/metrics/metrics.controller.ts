import { Controller, Get, Inject } from '@tsed/common';
import { ContentType, Description, Returns, Summary } from '@tsed/schema';
import { ProtocolMetricModel } from './interfaces/protocol-metric-model';
import { MetricsService } from './metrics.service';

@Controller('/metrics')
export class MetricsController {
  @Inject()
  metricsService!: MetricsService;

  @Get('')
  @ContentType('json')
  @Summary('Get a metric of the protocol across all chains')
  @Description('Returns the total amount of users, total amount of vaults and total value locked across all chains')
  @Returns(200, ProtocolMetricModel)
  @(Returns(404).Description('Protocol metrics not available'))
  async getProtocolMetrics() {
    return this.metricsService.getProtocolMetrics();
  }
}
