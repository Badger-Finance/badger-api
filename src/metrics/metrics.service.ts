import { Service } from '@tsed/common';
import { NotFound } from '@tsed/exceptions';

import { getDataMapper } from '../aws/dynamodb.utils';
import { ProtocolMetricSnapshot } from '../aws/models/protocol-metric-snapshot.model';
import { MetricType } from './enums/metric-type';
import { ProtocolMetrics } from './interfaces/metrics.interface';

@Service()
export class MetricsService {
  async getProtocolMetrics(): Promise<ProtocolMetrics> {
    const mapper = getDataMapper();
    let result: ProtocolMetrics | undefined;

    try {
      for await (const metric of mapper.query(
        ProtocolMetricSnapshot,
        { type: MetricType.protocol },
        { scanIndexForward: false, limit: 1 },
      )) {
        result = metric;
      }
    } catch (error) {
      console.error(error);
      throw new NotFound('Protocol metrics not available');
    }

    if (!result) {
      throw new NotFound('Protocol metrics not available');
    }

    return result;
  }
}
