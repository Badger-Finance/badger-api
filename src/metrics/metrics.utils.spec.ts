import { VaultDTO } from '@badger-dao/sdk';

import * as accountsUtils from '../accounts/accounts.utils';
import { VaultDefinitionModel } from '../aws/models/vault-definition.model';
import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import { mockChainVaults } from '../test/tests.utils';
import { fullTokenMockMap } from '../tokens/mocks/full-token.mock';
import * as tokenUtils from '../tokens/tokens.utils';
import * as vaultUtils from '../vaults/vaults.utils';
import { getProtocolMetrics, getProtocolSettMetrics, getProtocolTotalUsers } from './metrics.utils';

describe('metrics.utils', () => {
  beforeEach(() => {
    mockChainVaults();
    jest
      .spyOn(accountsUtils, 'getAccounts')
      .mockReturnValue(
        Promise.resolve([
          '0x0000000000000000000000000000000000000000',
          '0x0000000000000000000000000000000000000001',
          '0x0000000000000000000000000000000000000002',
          '0x0000000000000000000000000000000000000003',
          '0x0000000000000000000000000000000000000004',
          '0x0000000000000000000000000000000000000005',
          '0x0000000000000000000000000000000000000006',
        ]),
      );
    jest.spyOn(tokenUtils, 'getFullToken').mockImplementation(async (_, tokenAddr) => {
      return fullTokenMockMap[tokenAddr] || fullTokenMockMap[TOKENS.BADGER];
    });
    jest
      .spyOn(vaultUtils, 'getCachedVault')
      .mockImplementation(
        async (chain: Chain, VaultDefinition: VaultDefinitionModel): Promise<VaultDTO> =>
          vaultUtils.defaultVault(chain, VaultDefinition),
      );
  });

  describe('getProtocolUsersMetric', () => {
    it('returns total users', async () => {
      const totalUsers = await getProtocolTotalUsers();
      expect(totalUsers).toMatchSnapshot();
    });
  });

  describe('getSettsMetrics', () => {
    it('returns setts metrics', async () => {
      const metrics = await getProtocolSettMetrics();
      expect(metrics).toMatchSnapshot();
    });
  });

  describe('getProtocolMetrics', () => {
    it('returns protocolMetrics', async () => {
      const metrics = await getProtocolMetrics();
      expect(metrics).toMatchSnapshot();
    });
  });
});
