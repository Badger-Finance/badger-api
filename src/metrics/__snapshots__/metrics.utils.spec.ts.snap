// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metrics.utils getProtocolMetrics returns protocolMetrics 1`] = `
Object {
  "totalUsers": 7,
  "totalValueLocked": 0,
  "totalVaults": 61,
}
`;

exports[`metrics.utils getProtocolUsersMetric returns total users 1`] = `7`;

exports[`metrics.utils getSettsMetrics returns setts metrics 1`] = `
Object {
  "totalValueLocked": 0,
  "totalVaults": 61,
}
`;
